### PR TITLE
Fixed #29979, Refs #17337 -- Extracted AutoField field logic into a mixin and refactored AutoFields.

### PR DESCRIPTION
**Pull request by pope1ni**
Ticket [#29979](https://code.djangoproject.com/ticket/29979).

The first two commits are minor tweaks noticed while developing this set of changes. The last three can be squashed together, but were left separate for reviewing purposes.

### CODE REVIEW

File: django/db/backends/base/operations.py [hunk 1]
```diff
@@ -611,7 +611,8 @@ def integer_field_range(self, internal_type):
         return a tuple of the (min_value, max_value) form representing the
         range of the column type bound to the field.
         """
-        return self.integer_field_ranges[internal_type]
+        override = {'AutoField': 'IntegerField', 'BigAutoField': 'BigIntegerField'}
```
[the remark this answers is not in the document]
Comment (pope1ni) [reply django/db/backends/base/operations.py]: Hmm. The PostgreSQL backend uses `serial` and `bigserial` for `AutoField` and `BigAutoField` so the minimum should be `1` according to the [documentation](https://www.postgresql.org/docs/current/datatype-numeric.html#DATATYPE-NUMERIC-TABLE). In theory this ought to imply that `allows_auto_pk_0 = False` should be set for PostgreSQL, but no tests currently seem to be failing...

**Comment by charettes** [comment]
Hello @ngnpope!

Any thoughts on using a `AutoField(IntegerField)` and `BigAutoField(AutoField, BigIntegerField)` inheritance chain and getting rid of `AutoFieldMixin` instead?

That would do the same job at removing the code duplication without breaking the `BigAutoField(AutoField)` inheritance assumption which required the `isinstance(foo, AutoFieldMixin)` adjustment you made in core and are probably breaking a few third party apps as well.

**Comment by charettes** [comment]
I see where you are coming from. My main concerns with introducing `AutoFieldMixin` now without having a proper non-integer auto-field to use it is that we'll probably be missing something and need to go back to adjust it once we land support for generic auto-fields. That might means breaking the now _public_ `AutoFieldMixin` API.

Without testing a non-integer implementation we're not asserting the mixin is generic enough to be used for such fields such as `UUIDField`, we only provide a mixin we know is working for integer based fields.

**Comment by pope1ni** [comment]
That's fair. Will revisit this next week.

**Comment by pope1ni** [comment]
Will continue this for 3.0, rebased.

File: django/db/models/fields/__init__.py [hunk 2]
```diff
@@ -2335,3 +2245,70 @@ def formfield(self, **kwargs):
...
+            value = self.get_prep_value(value)
+            value = connection.ops.validate_autopk_value(value)
+        return value
+
+    def get_prep_value(self, value):
+        from django.db.models.expressions import OuterRef
+        return value if isinstance(value, OuterRef) else super().get_prep_value(value)
+
+    def contribute_to_class(self, cls, name, **kwargs):
+        assert not cls._meta.auto_field, "Model %s can't have more than one AutoField." % cls._meta.label
```
[the remark this answers is not in the document]
Comment (auvipy) [reply django/db/models/fields/__init__.py]: more than one automatically generated field.? sounds better and more natural with the changes.

**Comment by felixxm** [comment]
@pope1ni Is this ready for the another round of reviews? if yes please uncheck `Patch needs improvement` on the ticket.

**Comment by pope1ni** [comment]
@felixxm Not quite yet, but have been meaning to get back to it soon.

**Comment by pope1ni** [comment]
@felixxm @charettes

So I've had a bit more of a play with this and it is not possible to subclass both ``BigIntegerField`` and ``AutoField`` in ``BigAutoField`` as that screws up MRO as there are two paths to ``Field``. This means the only solution is to create a metaclass defining ``__instancecheck__`` and ``__subclasscheck__`` which seems to work well. These hooks could possibly be used later to add deprecation for inheritance from ``AutoField`` when and if ``AutoFieldMixin`` is public API, and also deprecation of ``isinstance(..., AutoField)`` checks when [#24042](https://code.djangoproject.com/ticket/24042) has been addressed. (As such I have reverted from using ``isinstance(..., AutoFieldMixin)`` to keep ``AutoFieldMixin`` private for now.)

I had a look at whether it would be possible to create a custom, non-integer ``AutoField``, e.g. ``UUIDAutoField``, but I think that this is not currently possible without support for database defaults, see [#470](https://code.djangoproject.com/ticket/470). All the sequence/identity/serial stuff that generates a value is very much integer only. However, this work to resolve [#29979](https://code.djangoproject.com/ticket/29979) does actually partially address some of [#17337](https://code.djangoproject.com/ticket/17337) as the new mixin has no reference to anything integer related.

While I was playing around with all of this, I added support for ``SmallAutoField`` to complete the set and to verify that this all works as expected, see [#30661](https://code.djangoproject.com/ticket/30661).

I think this is now ready for review.

**Comment by pope1ni** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
@pope1ni Can we move `SmallAutoField` to a separate PR? :pray:  It seems that it can be easily added in advance (without this refactoring).

**Comment by pope1ni** [comment]
@felixxm This includes the commit from #11611 as it now builds on top of that. Once that is merged, it should be simple to rebase this off master and remove the first commit.

**Comment by pope1ni** [comment]
buildbot, test on oracle.

**Comment by pope1ni** [comment]
Test failure is unrelated.

**Comment by pope1ni** [comment]
buildbot, test on oracle.

**Comment by pope1ni** [comment]
@felixxm I have rebased this now that ``SmallAutoField`` has been merged.

File: django/db/backends/base/operations.py [hunk 3]
```diff
@@ -611,7 +611,12 @@ def integer_field_range(self, internal_type):
         return a tuple of the (min_value, max_value) form representing the
         range of the column type bound to the field.
         """
-        return self.integer_field_ranges[internal_type]
+        override = {
+            'AutoField': 'IntegerField',
+            'BigAutoField': 'BigIntegerField',
+            'SmallAutoField': 'SmallIntegerField',
```
Comment (felixxm) [inline django/db/backends/base/operations.py]: I think we should add `SmallAutoField: (-99999, 99999)` to the `DatabaseOperations.integer_field_ranges` in the Oracle back-end.

Comment (pope1ni) [reply django/db/backends/base/operations.py]: I put in this override thing to avoid having to update `integer_field_ranges` in all of the backends, but if any custom backends override `.integer_field_range()` without calling `super()` this will break anyway...

So I'll just update `integer_field_ranges` in all of the backends and update the release notes to state that third-party backends need to update `integer_field_ranges` for the three auto fields also...

Comment (felixxm) [reply django/db/backends/base/operations.py]: Maybe:
```python
if internal_type in self.integer_field_ranges:
    return self.integer_field_ranges[internal_type]
return self.integer_field_ranges[override.get(internal_type)]
```

**Comment by pope1ni** [comment]
buildbot, test on oracle.

File: django/db/backends/base/operations.py (hunk 3 above)
Comment (pope1ni) [reply django/db/backends/base/operations.py]: I opted for not overriding `.integer_field_range()` in favour of this which is much less of a kludge:
```diff
diff --git a/django/db/backends/base/operations.py b/django/db/backends/base/operations.py
index d17dc8dcce..76abc6dcb2 100644
--- a/django/db/backends/base/operations.py
+++ b/django/db/backends/base/operations.py
@@ -26,6 +26,9 @@ class BaseDatabaseOperations:
         'BigIntegerField': (-9223372036854775808, 9223372036854775807),
         'PositiveSmallIntegerField': (0, 32767),
         'PositiveIntegerField': (0, 2147483647),
+        'SmallAutoField': (-32768, 32767),
+        'AutoField': (-2147483648, 2147483647),
+        'BigAutoField': (-9223372036854775808, 9223372036854775807),
     }
     set_operators = {
         'union': 'UNION',
diff --git a/django/db/backends/oracle/operations.py b/django/db/backends/oracle/operations.py
index 016d4c9a59..3cad02fa41 100644
--- a/django/db/backends/oracle/operations.py
+++ b/django/db/backends/oracle/operations.py
@@ -16,13 +16,18 @@ from .utils import BulkInsertMapper, InsertVar, Oracle_datetime
 
 
 class DatabaseOperations(BaseDatabaseOperations):
-    # Oracle uses NUMBER(11) and NUMBER(19) for integer fields.
+    # Oracle uses NUMBER(5), NUMBER(11), and NUMBER(19) for integer fields.
+    # SmallIntegerField uses NUMBER(11) instead of NUMBER(5), which is used by
+    # SmallAutoField, to preserve backward compatibility.
     integer_field_ranges = {
         'SmallIntegerField': (-99999999999, 99999999999),
         'IntegerField': (-99999999999, 99999999999),
         'BigIntegerField': (-9999999999999999999, 9999999999999999999),
         'PositiveSmallIntegerField': (0, 99999999999),
         'PositiveIntegerField': (0, 99999999999),
+        'SmallAutoField': (-99999, 99999),
+        'AutoField': (-99999999999, 99999999999),
+        'BigAutoField': (-9999999999999999999, 9999999999999999999),
     }
     set_operators = {**BaseDatabaseOperations.set_operators, 'difference': 'MINUS'}
 
--- a/docs/releases/3.0.txt
+++ b/docs/releases/3.0.txt
@@ -392,6 +396,11 @@ backends.
 * ``DatabaseOperations.return_insert_id()`` now requires an additional
   ``field`` argument with the model field.
 
+* ``DatabaseOperations.integer_field_ranges`` has been updated to add entries
+  for ``AutoField``, ``BigAutoField``, and ``SmallAutoField`` to support the
+  standard integer range validators on these field types. Third-party backends
+  may need to override the default entries.
+
 :mod:`django.contrib.admin`
 ---------------------------
 
```

**Comment by pope1ni** [comment]
buildbot, test on oracle.

**Comment by pope1ni** [comment]
Rebased and made changes to address the valid ranges on Oracle.

Just trying to get the tests to run again as the last encountered unrelated errors:
```
django.db.utils.DatabaseError: ORA-01940: cannot drop a user that is currently connected
```
```
ModuleNotFoundError: No module named 'memcache'
```

**Comment by pope1ni** [comment]
@felixxm I believe this one is ready for final review.

**Comment by felixxm** [comment]
@pope1ni Thanks, it's on my list.

**Comment by pope1ni** [comment]
buildbot, test on oracle.

File: django/db/models/fields/__init__.py [hunk 4]
```diff
@@ -898,110 +898,6 @@ def value_from_object(self, obj):
...
-                self.error_messages['invalid'],
-                code='invalid',
-                params={'value': value},
-            )
-
-    def rel_db_type(self, connection):
-        return IntegerField().db_type(connection=connection)
-
-    def validate(self, value, model_instance):
-        pass
```
Comment (felixxm) [inline django/db/models/fields/__init__.py]: `AutoFieldMixin.validate()`  is missing , I think we should restore `AutoField.validate()` implementation, i.e.
```python
def validate(self, value, model_instance):
    pass
```

File: django/db/models/query.py [hunk 5]
```diff
@@ -15,11 +15,10 @@
     DJANGO_VERSION_PICKLE_KEY, IntegrityError, connections, router,
     transaction,
 )
-from django.db.models import DateField, DateTimeField, sql
+from django.db.models import AutoField, DateField, DateTimeField, sql
 from django.db.models.constants import LOOKUP_SEP
 from django.db.models.deletion import Collector
 from django.db.models.expressions import Case, Expression, F, Value, When
-from django.db.models.fields import AutoField
```
Comment (felixxm) [inline django/db/models/query.py]: This change is not related with the ticket. Please revert it, we can make this cleanup (and probably similar imports cleanups) in a separate PR.

Comment (pope1ni) [reply django/db/models/query.py]: That's fair. Will revert and raise a separate PR.

File: django/db/models/fields/__init__.py (hunk 4 above)
Comment (pope1ni) [reply django/db/models/fields/__init__.py]: Interesting. It has been like this since 471596fc1afcb9c6258d317c619eaf5fd394e797, but I don't really understand the reason that it is bypassing [`Field.validate()`](471596fc1afcb9c6258d317c619eaf5fd394e797), unless it is a minor optimisation to skip it because none of those checks really make sense for the auto fields...

Will restore this.

**Comment by pope1ni** [comment]
buildbot, test on oracle.

File: django/db/models/fields/__init__.py [hunk 6]
```diff
@@ -2395,3 +2291,113 @@ def formfield(self, **kwargs):
...
+            value = connection.ops.validate_autopk_value(value)
+        return value
+
+    def get_prep_value(self, value):
+        from django.db.models.expressions import OuterRef
+        return value if isinstance(value, OuterRef) else super().get_prep_value(value)
+
+    def contribute_to_class(self, cls, name, **kwargs):
+        assert not cls._meta.auto_field, (
+            "Model %s can't have more than one auto-incrementing field."
```
Comment (pope1ni) [inline django/db/models/fields/__init__.py]: So `auto-incrementing` implies that this is numeric. Given we're aiming toward non-integer support, do we want to rephrase this slightly to avoid needing to change it again later? I'm conscious that support for generated columns is also on the horizon, so we need to be careful with using the word "generated" here too, but "auto-generated" _may_ be OK...

I'll leave this to you to change if you see fit prior to merge.

Comment (felixxm) [reply django/db/models/fields/__init__.py]: I don't think that we won't have this restriction anymore for non-integer auto-field (at least not in all databases because PostgreSQL and Oracle support `RETURNING` clause with multiple fields). Let's leave `auto-generated` for now.

**Comment by pope1ni** [comment]
Thanks @felixxm! :star2:

File: django/db/models/query.py (hunk 5 above)
Comment (blueyed) [reply django/db/models/query.py]: For reference: https://github.com/django/django/pull/11689
